src/track/trackref.cpp

```diff
@@ -0,0 +1,43 @@
+#include "track/trackref.h"
+
+
+bool TrackRef::verifyConsistency() const {
+    DEBUG_ASSERT_AND_HANDLE(hasLocation() || !hasCanonicalLocation()) {
+        return false;
+    }
+    DEBUG_ASSERT_AND_HANDLE(hasLocation() || !hasId()) {
+        return false;
+    }
+    return true;
+}
+
+QFileInfo TrackRef::createFileInfo() const {
+    QFileInfo fileInfo(getLocation());
```

Comment:
TrackRef is supposed to be a copyable value object. Therefore adding a mutex 
to it is not an option.

By tracing the invocations of createFileInfo() we can find a place where we 
could to keep or cache the created QFileInfo instance. If we are accessing 
the actual file anyway the overhead of creating a temporary QFileInfo 
instance should not matter.

On 09/17/2015 12:50 PM, Daniel Schürmann wrote:
>
> In src/track/trackref.cpp 
> <https://github.com/mixxxdj/mixxx/pull/709#discussion_r39732614>:
>
> > @@ -0,0 +1,43 @@
> > +#include "track/trackref.h"
> > +
> > +
> > +bool TrackRef::verifyConsistency() const {
> > +    DEBUG_ASSERT_AND_HANDLE(hasLocation() || !hasCanonicalLocation()) {
> > +        return false;
> > +    }
> > +    DEBUG_ASSERT_AND_HANDLE(hasLocation() || !hasId()) {
> > +        return false;
> > +    }
> > +    return true;
> > +}
> > +
> > +QFileInfo TrackRef::createFileInfo() const {
> > +    QFileInfo fileInfo(getLocation());
>
> I have read a bit in the source and docs of QFileInfo and it turns out 
> that it is very optimized.
> It does only things that are required to fulfill a query, but has a cache 
> for results of system calls.
> The only mayor overhead in the constructor is checking for "\" or "/", and 
> preparing the underlying FileEngine.
>
> If we create temporary file Infos QFileInfo by calling createFileInfo() 
> over and over again, we do not benefit from the caching nature of 
> QFileInfo but we have always the penalty of constructing a new QFileInfo.
> However QFileInfo is not thread save.
>
> So we have on one other hand the locking inside Mixxx, to system calls, 
> that are probably locking anyway,
> on the other hand we introduce CPU overhead by constructing QFileInfo.
> If we consider that the locks are never locked, because a TrackPointer is 
> used only at on thread at a time, it is a performance regression.
>
> How about that: Add a QFileInfo member to TrackRef and reintroduce a mutex 
> for it.
> Use it for all calls that require a system call, which is probably locking 
> anyway.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/709/files#r39732614>.
>